### ISSUE TITLE
After training on a new dataset of images the model is later on also detecting objects it was trained by default (laptop, cup etc.).

### ISSUE THREAD

**Issue by HemanParbhakar** [tensorflow/models#9295]
# Prerequisites

Please answer the following questions for yourself before submitting an issue.

- [ ] I am using the latest TensorFlow Model Garden release and TensorFlow 2.
- [ ] I am reporting the issue to the correct repository. (Model Garden official or research directory)
- [ ] I checked to make sure that this issue has not already been filed.

## 1. The entire URL of the file you are using

https://github.com/tensorflow/models/tree/master/research/...

## 3. Steps to reproduce

Steps to reproduce the behavior.

## 5. Additional context

Include any logs that would be helpful to diagnose the problem.

## 6. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- Mobile device name if the issue happens on a mobile device:
- TensorFlow installed from (source or binary):
- TensorFlow version (use command below):
- Python version:
- Bazel version (if compiling from source):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**Comment by saikumarchalla** [comment]
@HemanParbhakar  Could  you please fill the issue Template with proper information. Also, Let us know which pre-trained model are you referring to?. Thanks!

**Comment by HemanParbhakar** [comment]
The model is ssd 320x320 object detection. Trained on a custom data set.

**Comment by gowthamkpr** [comment]
@HemanParbhakar Can you please explain what issue you are facing here. Thanks!

**Comment by HemanParbhakar** [comment]
@gowthamkpr i trained a model using customer data set (e.g cat). After training is done model is converted and implemented in the application. Then model detects laptop, remote, water botel as cat. Why is this happening if transfer learning is done using images of cat.

**Comment by HemanParbhakar** [comment]
any update?

**Comment by gowthamkpr** [comment]
@HemanParbhakar Before converting it to tflite, ate the predictions of the model accurate?

**Comment by HemanParbhakar** [comment]
@srjoglekar246 i have done the same way. as you mentioned in of the issues as well.

**Comment by HemanParbhakar** [comment]
> The model is ssd 320x320 object detection. Trained on a custom data set.
> https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/tf2_detection_zoo.md
> Training is done
> https://tensorflow-object-detection-api-tutorial.readthedocs.io/en/latest/training.html#training-the-model
> Conversion:
> https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/running_on_mobile_tf2.md
> 
> The ssd_max_detections is set to 100 and ssd_use_regular_nms to true to increase detection.

I have followed this processes

**Comment by srjoglekar246** [comment]
None of those links describe how to *run* the model, as in Invoke() it, on-device.

**Comment by HemanParbhakar** [comment]
I am doing that in Android

**Comment by srjoglekar246** [comment]
Were you able to verify correctness of the TF version of the model? What metrics are you using?

Which label_map are you using for your Android inference? How many output classes does your SSD model have? And how are you providing the input tensor & interpreting the output tensors of the model?

Unfortunately, "in Android" is pretty vague to debug what might be going wrong in running the model on device :-)

**Comment by HemanParbhakar** [comment]
I have done my training on custom data set which has custom label_map file. The Output class you mean in pipeline.config(It was 1 and if you saying while converting there i have given 100 as you said to increase accuracy).  
The input to model was given as tfrecord

-

**Comment by srjoglekar246** [comment]
If the number of output classes is only 1, how can the model provide multiple labels as output?
And how are you providing tfrecord as input to TFLite on Android? That isn't supported.


@sayakpaul has some experience with these steps, can you take a look?

**Comment by HemanParbhakar** [comment]
the input is tensor in Android.
The label is providing a single output only.
But suppose the model is trained to detect dogs and when the model is trained and added in the application it shows laptop as dogs as well.

**Comment by srjoglekar246** [comment]
What is the output score of that detection? And its really difficult to debug this unless you provide the *code* you are using the run this model on Android.

**Comment by HemanParbhakar** [comment]
but @srjoglekar246 The code is same as tensorflow example no changes except model,label file . The score was 87%

**Comment by srjoglekar246** [comment]
Did you change [these params](https://github.com/tensorflow/examples/blob/master/lite/examples/object_detection/android/app/src/main/java/org/tensorflow/lite/examples/detection/DetectorActivity.java#L54) based on the model?

**Comment by HemanParbhakar** [comment]
yes

**Comment by srjoglekar246** [comment]
How did you verify correctness of the TF model?

**Comment by HemanParbhakar** [comment]
I have tested on Dog it detects the dog properly but why does it says laptop, mobile is also a dog

**Comment by sayakpaul** [comment]
If you ensure the correctness of the TensorFlow Lite model locally i.e. with a Python script referring to the Colab @srjoglekar246 mentioned, there is probably something wrong with the way you parsing the outputs from your mobile application.

**Comment by HemanParbhakar** [comment]
I have tested the model locally it works well on the system but on the device it is causing the issue.

**Comment by HemanParbhakar** [comment]
> If you ensure the correctness of the TensorFlow Lite model locally i.e. with a Python script referring to the Colab @srjoglekar246 mentioned, there is probably something wrong with the way you parsing the outputs from your mobile application.

But i am using https://github.com/tensorflow/examples/tree/master/lite/examples/object_detection/android repo for testing my model only changed some parameter nothing else.

**Comment by HemanParbhakar** [comment]
@srjoglekar246 any update?

**Comment by srjoglekar246** [comment]
@HemanParbhakar It might be better to check in Python first, as a sanity check to see if the model is at all correct. This will also help us figure out if the problem is with the model, or with the android app inference code.

**Comment by HemanParbhakar** [comment]
> @HemanParbhakar It might be better to check in Python first, as a sanity check to see if the model is at all correct. This will also help us figure out if the problem is with the model, or with the android app inference code.

But the inference is done on the sample application provided by tensorflow and inference on the code you provided.
